v2/cf-nodejs-logging-support/src/lib/logger/record-factory.ts

```diff
@@ -62,56 +88,125 @@ export default class RecordFactory {
+    private prepareStacktrace(stacktraceStr: any): any {
+        var fullStacktrace = stacktraceStr.split('\n');
+        var totalLineLength = fullStacktrace.reduce((acc: any, line: any) => acc + line.length, 0);
+
+        if (totalLineLength > this.MAX_STACKTRACE_SIZE) {
+            var truncatedStacktrace = [];
+            var stackA = [];
+            var stackB = [];
+            var indexA = 0;
+            var indexB = fullStacktrace.length - 1;
+            var currentLength = 73; // set to approx. character count for "truncated" and "omitted" labels
+
+            for (let i = 0; i < fullStacktrace.length; i++) {
+                if (i % 3 == 0) {
+                    let line = fullStacktrace[indexA++];
+                    if (currentLength + line.length > this.MAX_STACKTRACE_SIZE) {
+                        break;
                     }
-                    return process.env[field.source.name!];
-                case "config-field":
-                    return record[field.source.name!];
-                default:
-                    return undefined;
+                    currentLength += line.length;
+                    stackA.push(line);
+                } else {
+                    let line = fullStacktrace[indexB--];
+                    if (currentLength + line.length > this.MAX_STACKTRACE_SIZE) {
+                        break;
+                    }
+                    currentLength += line.length;
+                    stackB.push(line);
+                }
             }
-        } else {
 
-            // TO DO: handle sources as array case
+            truncatedStacktrace.push("-------- STACK TRACE TRUNCATED --------");
+            truncatedStacktrace = [...truncatedStacktrace, ...stackA];
+            truncatedStacktrace.push(`-------- OMITTED ${fullStacktrace.length - (stackA.length + stackB.length)} LINES --------`);
+            truncatedStacktrace = [...truncatedStacktrace, ...stackB.reverse()];
+            return truncatedStacktrace;
         }
+        return fullStacktrace;
+    }
 
-        return;
+    private isValidObject(obj: any, canBeEmpty?: any): boolean {
+        if (!obj) {
+            return false;
+        } else if (typeof obj !== "object") {
+            return false;
+        } else if (!canBeEmpty && Object.keys(obj).length === 0) {
+            return false;
+        }
+        return true;
+    }
+
+    private addCustomFields(record: any, registeredCustomFields: Array<string>, loggerCustomFields: Map<string, any>, customFieldsFromArgs: any): any {
+        var providedFields = Object.assign({}, loggerCustomFields, customFieldsFromArgs);
+        const customFieldsFormat = Config.getInstance().getConfig().customFieldsFormat;
+        const isCloudLogging = customFieldsFormat == "cloud-logging";
+        const isAppLogging = customFieldsFormat == "application-logging";
```

Comment:
Please define "cloud-logging" and "application-logging" as constants instead of using constants for "isCloudLogging" and "isAppLogging".